File: flint/strategy/ma_crossover.py
"""Simple moving-average crossover strategy."""
from __future__ import annotations

from typing import TYPE_CHECKING, Dict, List, Optional

import numpy as np

from ..models import Candle, Signal
from .base import Strategy

if TYPE_CHECKING:
    from ..execution.context import ExecutionContext


class MACrossoverStrategy(Strategy):
    """Go long when fast SMA crosses above slow SMA, exit when it crosses below."""

    def __init__(self, fast_period: int = 10, slow_period: int = 30) -> None:
        if fast_period >= slow_period:
            raise ValueError("fast_period must be < slow_period")
        self.fast_period = fast_period
        self.slow_period = slow_period
        self._prev_fast: float = 0.0
        self._prev_slow: float = 0.0

    @property
    def name(self) -> str:
        return f"MA-Crossover({self.fast_period}/{self.slow_period})"

    def reset(self) -> None:
        self._prev_fast = 0.0
        self._prev_slow = 0.0

    @classmethod
    def parameters(cls) -> Dict[str, dict]:
        return {
            "fast_period": {"type": "int", "low": 5, "high": 50, "default": 10},
            "slow_period": {"type": "int", "low": 20, "high": 200, "default": 30},
        }

    def on_candle(self, candle: Candle, history: List[Candle], ctx: Optional["ExecutionContext"] = None) -> Signal:
        if len(history) < self.slow_period:
            return Signal.HOLD

        closes = np.array([c.close for c in history[-self.slow_period :]])
        fast_ma = float(np.mean(closes[-self.fast_period :]))
        slow_ma = float(np.mean(closes))

        signal = Signal.HOLD
        if self._prev_fast != 0.0 and self._prev_slow != 0.0:
            if self._prev_fast <= self._prev_slow and fast_ma > slow_ma:
                signal = Signal.BUY
            elif self._prev_fast >= self._prev_slow and fast_ma < slow_ma:
                signal = Signal.SELL

        self._prev_fast = fast_ma
        self._prev_slow = slow_ma
        return signal

File: ui/src/templates/ma-crossover.ts
export const MA_CROSSOVER_TEMPLATE = `from flint.strategy.base import Strategy
from flint.models import Candle, Signal, Side, TimeInForce
from typing import List
import numpy as np


class MACrossover(Strategy):
    """Moving Average Crossover — trend following.

    Buys when fast SMA crosses above slow SMA (golden cross).
    Sells when fast crosses below slow (death cross).
    Uses v2 execution: stop-loss, impact checks, position sizing.
    """
    def __init__(self, fast_period=10, slow_period=30, stop_pct=3.0):
        self.fast = fast_period
        self.slow = slow_period
        self.stop_pct = stop_pct / 100
        self._prev_fast = None
        self._prev_slow = None

    @property
    def name(self) -> str:
        return f"MA-Crossover({self.fast}/{self.slow})"

    @classmethod
    def parameters(cls):
        return {
            "fast_period": {"type": "int", "low": 5, "high": 50, "default": 10},
            "slow_period": {"type": "int", "low": 20, "high": 200, "default": 30},
            "stop_pct": {"type": "float", "low": 1.0, "high": 8.0, "default": 3.0},
        }

    def on_candle(self, candle: Candle, history: List[Candle], ctx=None) -> Signal:
        if len(history) < self.slow:
            return Signal.HOLD
        closes = [c.close for c in history[-self.slow:]]
        fast_ma = float(np.mean(closes[-self.fast:]))
        slow_ma = float(np.mean(closes))

        if ctx is None or self._prev_fast is None:
            self._prev_fast = fast_ma
            self._prev_slow = slow_ma
            return Signal.HOLD

        golden_cross = self._prev_fast <= self._prev_slow and fast_ma > slow_ma
        death_cross = self._prev_fast >= self._prev_slow and fast_ma < slow_ma
        self._prev_fast = fast_ma
        self._prev_slow = slow_ma

        if golden_cross and not ctx.positions:
            size = (ctx.account.cash * 0.9) / candle.close
            impact = ctx.get_impact_price(candle.market, Side.LONG, size)
            if impact and abs(impact - candle.close) / candle.close > 0.002:
                ctx.log(f"Skip entry: impact too high ({abs(impact - candle.close)/candle.close*10000:.0f}bps)")
                return Signal.HOLD
            if size > 0:
                ctx.market_order(candle.market, Side.LONG, size)
                ctx.stop_order(candle.market, Side.SHORT, size,
                               candle.close * (1 - self.stop_pct))

        elif death_cross and ctx.positions:
            ctx.close_position(candle.market)
            ctx.cancel_all(candle.market)

        return Signal.HOLD

    def reset(self) -> None:
        self._prev_fast = None
        self._prev_slow = None
`
